Mathlib/GroupTheory/SpecificGroups/Cyclic.lean

```diff
@@ -680,3 +680,90 @@ lemma not_isCyclic_iff_exponent_eq_prime [Group α] {p : ℕ} (hp : p.Prime)
+  let m : ℤ := n
+  change _ ∣ m ↔ ∃ k, k * (orderOf g : ℤ) = m
+  conv in _ * _ = _ => rw [mul_comm, eq_comm]
+
+/-- The kernel of `zmultiplesHom G g` is equal to the additive subgroup generated by
+    `addOrderOf g`. -/
+theorem zmultiplesHom_ker_eq [AddGroup G] (g : G) :
+    (zmultiplesHom G g).ker = zmultiples ↑(addOrderOf g) := by
+  ext n
+  rw [AddMonoidHom.mem_ker, mem_zmultiples_iff, zmultiplesHom_apply,
+    ← addOrderOf_dvd_iff_zsmul_eq_zero]
+  let m : ℤ := n
+  change _ ∣ m ↔ ∃ k, k * (addOrderOf g : ℤ) = m
+  conv in _ * _ = _ => rw [mul_comm, eq_comm]
+
+/-- The isomorphism from `ZMod n` to any cyclic additive group of `Nat.card` equal to `n`. -/
+noncomputable def zmodAddCyclicAddEquiv [AddGroup G] (h : IsAddCyclic G) :
+    ZMod (Nat.card G) ≃+ G := by
+  let n := Nat.card G
+  let ⟨g, surj⟩ := Classical.indefiniteDescription _ h.exists_generator
+  have kereq : ((zmultiplesHom G) g).ker = zmultiples ↑(Nat.card G) := by
+    rw [zmultiplesHom_ker_eq]
+    congr
+    rw [← Nat.card_zmultiples]
+    exact Nat.card_congr (Equiv.subtypeUnivEquiv surj)
+  exact (Int.quotientZmultiplesNatEquivZMod n)
+    |>.symm.trans <| QuotientAddGroup.quotientAddEquivOfEq kereq
+    |>.symm.trans <| QuotientAddGroup.quotientKerEquivOfSurjective (zmultiplesHom G g) surj
+
+/-- The isomorphism from `Multiplicative (ZMod n)` to any cyclic group of `Nat.card` equal to `n`.
+-/
+noncomputable def zmodCyclicMulEquiv [Group G] (h : IsCyclic G) :
+    Multiplicative (ZMod (Nat.card G)) ≃* G :=
+  AddEquiv.toMultiplicative <| zmodAddCyclicAddEquiv <| isAddCyclic_additive_iff.2 h
+
+/-- Two cyclic groups of the same cardinality are isomorphic. -/
+noncomputable def mulEquivOfCyclicCardEq [Group G] [Group H] (hG : IsCyclic G)
+    (hH : IsCyclic H) (hcard : Nat.card G = Nat.card H) : G ≃* H := hcard ▸
+  zmodCyclicMulEquiv hG |>.symm.trans (zmodCyclicMulEquiv hH)
+
+/-- Two cyclic additive groups of the same cardinality are isomorphic. -/
+noncomputable def addEquivofAddCyclicCardEq [AddGroup G] [AddGroup H] (hG : IsAddCyclic G)
+    (hH : IsAddCyclic H) (hcard : Nat.card G = Nat.card H) : G ≃+ H := hcard ▸
+  zmodAddCyclicAddEquiv hG |>.symm.trans (zmodAddCyclicAddEquiv hH)
+
+/-- Two groups of the same prime cardinality are isomorphic. -/
+noncomputable def mulEquivOfPrimeCardEq {p : ℕ} [Fintype G] [Fintype H] [Group G] [Group H]
+    [Fact p.Prime] (hG : Fintype.card G = p) (hH : Fintype.card H = p) : G ≃* H := by
+  apply mulEquivOfCyclicCardEq (isCyclic_of_prime_card hG) (isCyclic_of_prime_card hH)
+  rw [← Nat.card_eq_fintype_card] at hG hH
+  exact hG.trans hH.symm
+
+/-- Two additive groups of the same prime cardinality are isomorphic. -/
+noncomputable def addEquivOfPrimeCardEq {p : ℕ} [Fintype G] [Fintype H] [AddGroup G] [AddGroup H]
+    [Fact p.Prime] (hG : Fintype.card G = p) (hH : Fintype.card H = p) : G ≃+ H := by
+  apply addEquivofAddCyclicCardEq (isAddCyclic_of_prime_card hG) (isAddCyclic_of_prime_card hH)
```

Comment:
```suggestion
  have hGcyc := isAddCyclic_of_prime_card hG
  have hHcyc := isAddCyclic_of_prime_card hH
  apply addEquivofAddCyclicCardEq
```

Reply:
It seems like `@[to_additive]` has issues with everything in this section. For example:
```lean
/-
application type mismatch
  MonoidHom.ker
argument has type
  AddZeroClass G
but function has type
  [inst : MulOneClass G] → (Multiplicative ℤ →* G) → Subgroup (Multiplicative ℤ)
-/
@[to_additive test]
theorem zpowersHom_ker_eq [Group G] (g : G) :
    (zpowersHom G g).ker = zpowers (Multiplicative.ofAdd ↑(orderOf g)) := by
  ext n
  rw [MonoidHom.mem_ker, mem_zpowers_iff, zpowersHom_apply, ← orderOf_dvd_iff_zpow_eq_one]
  let m : ℤ := n
  change _ ∣ m ↔ ∃ k, k * (orderOf g : ℤ) = m
  conv in _ * _ = _ => rw [mul_comm, eq_comm]
```